mealie-recipes/mealie
# Feature: "I Made This" Dialog

## What type of PR is this?

_(REQUIRED)_



- feature

## What this PR does / why we need it:

_(REQUIRED)_

This adds a "Last Made" property to recipes and a new "I Made This" dialog on the recipe page. It creates a new event on the recipe event timeline and updates the recipe's `lastMade` timestamp to today (unless the user chooses a specific date).

The last made date is also displayed on the recipe page, and recipes can be sorted by this property.
![image](https://user-images.githubusercontent.com/71845777/199816797-3337631e-c4e7-4b2f-b245-8fc62d961b6c.png)
*Works on both landscape and non-landscape view, and does not appear on the public recipe page. The date format is localized*


![image](https://user-images.githubusercontent.com/71845777/199816557-489bc8e4-ae2c-41cc-9420-ff5f96e81067.png)
*If the recipe has never been made*


![image](https://user-images.githubusercontent.com/71845777/199816388-a49497d3-02a9-404c-a758-3a3cb3e745be.png)
*The comment isn't viewable in the UI yet, pending a recipe timeline view*


## Which issue(s) this PR fixes:

_(REQUIRED)_

Works towards #1451. This utilizes the timeline API, but does not implement the timeline frontend.

## Special notes for your reviewer:

_(fill-in or delete this section)_

Originally I had this in the recipe Action/Context menus, but I moved it to the recipe header/title instead since I think it works better there. It also made manipulating the last made date significantly easier.

I wasn't sure how to add new text localizations, so at the moment everything is just in English. If we want to add localizations in this PR I'm happy to do so, if you could help point me in the right direction. Not too much text was added, and it was only added in two places (the new component and the sort menu) so it should be pretty easy.

Also, to avoid the plugin conflict we talked about on discord, I modified the timeline event API to use `event_message` instead of `message`. This ended up being the easiest solution by far, since the plugin directly accesses the `POST` response. I also needed a new database migration for this PR anyway.

## Testing

_(fill-in or delete this section)_

I ran through the following scenarios:
1. Verify the dialog reactively updates the "last made" date without a refresh
2. Verify if the last made date is updated, then the user edits and saves the recipe without leaving the page, then the last made date is still the new date and not the old one (i.e. checking for the issue in #1397)
3. Verify the dialog can't be accessed when the recipe is being viewed publicly (i.e. there is no user)

## Release Notes

_(REQUIRED)_


```release-note
added tracking for when a recipe was last made ("I made this")
added sorting recipes by last made date
```
